database.go

```diff
@@ -0,0 +1,723 @@
+package influxdb
+
+import (
+	"encoding/json"
+	"regexp"
+	"sort"
+	"strings"
+	"time"
+)
+
+// database is a collection of retention policies and shards. It also has methods
+// for keeping an in memory index of all the measurements, series, and tags in the database.
+// Methods on this struct aren't goroutine safe. They assume that the server is handling
+// any locking to make things safe.
+type database struct {
+	name string
+
+	policies map[string]*RetentionPolicy // retention policies by name
+	shards   map[uint64]*Shard           // shards by id
+
+	defaultRetentionPolicy string
+
+	// in memory indexing structures
+	measurements        map[string]*Measurement // measurement name to object and index
+	seriesToMeasurement map[uint32]*Measurement // map series id to its measurement
```

Comment:
I haven't fully reviewed this code, but it occurs to me -- it you added the `Measurement` to the `Series` struct, perhaps you wouldn't need this map? To get a `Measurement` from a Series ID, could you then just do:
```
measurement := series[id].Measurement
```
1 less map? This assumes you have access to Measurement when creating the Series (I don't see why not).